lib/Conversion/StandardToStaticLogic/StandardToStaticLogic.cpp

```diff
@@ -0,0 +1,271 @@
+          // Add only unique uses
+          if (std::find(uses[&block].begin(), uses[&block].end(),
+                        op.getOperand(i)) == uses[&block].end())
+            uses[&block].push_back(op.getOperand(i));
+      }
+    }
+  }
+  return uses;
+}
+
+static BlockValues getBlockDefs(mlir::FuncOp f) {
+  // Returns map of values defined in each block
+  // For liveness analysis
+  BlockValues defs;
+
+  for (Block &block : f) {
+
+    // Values produced by operations in b
+    for (Operation &op : block) {
+      if (op.getNumResults() > 0) {
+        for (auto result : op.getResults())
+          defs[&block].push_back(result);
+      }
+    }
+
+    // Argument values of b
+    for (auto &arg : block.getArguments())
+      defs[&block].push_back(arg);
+  }
+
+  return defs;
+}
+
+static vector<Value> vectorUnion(vector<Value> v1, vector<Value> v2) {
+  // Returns v1 U v2
+  // Assumes unique values in v1
+
+  for (int i = 0, e = v2.size(); i < e; ++i) {
+    Value val = v2[i];
+    if (std::find(v1.begin(), v1.end(), val) == v1.end())
+      v1.push_back(val);
+  }
+  return v1;
+}
+
+static vector<Value> vectorDiff(vector<Value> v1, vector<Value> v2) {
+  // Returns v1 - v2
+  vector<Value> d;
+
+  for (int i = 0, e = v1.size(); i < e; ++i) {
+    Value val = v1[i];
+    if (std::find(v2.begin(), v2.end(), val) == v2.end())
+      d.push_back(val);
+  }
+  return d;
+}
+
+static BlockValues getBlockLiveIns(mlir::FuncOp f) {
+
+  // blockUses: values used in block but not defined in block
+  BlockValues blockUses = getBlockUses(f);
+
+  // blockDefs: values defined in block
+  BlockValues blockDefs = getBlockDefs(f);
+
+  BlockValues blockLiveIns;
+
+  bool change = true;
+  // Iterate while there are any changes to any of the livein sets
+  while (change) {
+    change = false;
+
+    // liveOuts(b): all liveins of all successors of b
+    // liveOuts(b) = U (blockLiveIns(s)) forall successors s of b
+    for (Block &block : f) {
+      vector<Value> liveOuts;
+      for (int i = 0, e = block.getNumSuccessors(); i < e; ++i) {
+        Block *succ = block.getSuccessor(i);
+        liveOuts = vectorUnion(liveOuts, blockLiveIns[succ]);
+      }
+
+      // diff(b):  liveouts of b which are not defined in b
+      // diff(b) = liveOuts(b) - blockDefs(b)
+      vector<Value> diff = vectorDiff(liveOuts, blockDefs[&block]);
+
+      // liveIns(b) = blockUses(b) U diff(b)
+      vector<Value> tmpLiveIns = vectorUnion(blockUses[&block], diff);
+
+      // Update blockLiveIns if new liveins found
+      if (tmpLiveIns.size() > blockLiveIns[&block].size()) {
+        blockLiveIns[&block] = tmpLiveIns;
+        change = true;
+      }
+    }
+  }
+  return blockLiveIns;
+}
+
+/// This function converts all live-ins of a block into its arguments, and
+/// converts all live-outs of a block into its return values. In this way, the
+/// sequencial code inside of each block can be isolated and converted to a
+/// function call.
```

Comment:
This seems similar to the live variables analysis in StandardToHandshake.  Perhaps this should get factored out rather than duplicating it again?  Hmm... is this even required or can we just leave the flow between pipeline stages implicit as long as the blocks are a linear chain?

Reply:
Yes, we can definitely factor this out as kind of canonicalization.

Since we will not isolate the pipeline operation as a function-like region, this analysis is no more necessary. We can just leave them implicit and all these stuff will happen automatically in the standard-to-handshake pass.